ambry-messageformat/src/main/java/com.github.ambry.messageformat/MessageFormatRecord.java

```diff
@@ -362,6 +399,192 @@ private void verifyCrc() throws MessageFormatException {
+   *                    user metadata size + blob size
+   *
+   *  Blob Encryption - The offset at which the blob encryption key record is located relative to this message.
+   *  Key relative      Non-existence of blob key record is indicated by -1. Blob Keys are optionally present for Put
+   *  offset            records. Blob Keys will be absent for Delete records.
+   *
+   *  blob property   - The offset at which the blob property record is located relative to this message. Only one of
+   *  relative offset   blob property/delete relative offset field can exist. Non existence is indicated by -1
+   *
+   *  delete          - The offset at which the delete record is located relative to this message. Only one of blob
+   *  relative offset   property/delete relative offset field can exist. Non existence is indicated by -1
+   *
+   *  user metadata   - The offset at which the user metadata record is located relative to this message. This exist
+   *  relative offset   only when blob property record and blob record exist
+   *
+   *  blob metadata   - The offset at which the blob record is located relative to this message. This exist only when
+   *  relative offset   blob property record and user metadata record exist
+   *
+   *  crc             - The crc of the message header
+   *
+   */
+  public static class MessageHeader_Format_V2 {
+    private ByteBuffer buffer;
+
+    // total size field start offset and size
+    public static final int Total_Size_Field_Offset_In_Bytes = Version_Field_Size_In_Bytes;
+    public static final int Total_Size_Field_Size_In_Bytes = 8;
+
+    // relative offset fields start offset and size
+    private static final int Number_Of_Relative_Offset_Fields = 5;
+    public static final int Relative_Offset_Field_Sizes_In_Bytes = 4;
+    public static final int Blob_Encryption_Key_Relative_Offset_Field_Offset_In_Bytes =
+        Total_Size_Field_Offset_In_Bytes + Total_Size_Field_Size_In_Bytes;
+    public static final int BlobProperties_Relative_Offset_Field_Offset_In_Bytes =
+        Blob_Encryption_Key_Relative_Offset_Field_Offset_In_Bytes + Relative_Offset_Field_Sizes_In_Bytes;
+    public static final int Delete_Relative_Offset_Field_Offset_In_Bytes =
+        BlobProperties_Relative_Offset_Field_Offset_In_Bytes + Relative_Offset_Field_Sizes_In_Bytes;
+    public static final int UserMetadata_Relative_Offset_Field_Offset_In_Bytes =
+        Delete_Relative_Offset_Field_Offset_In_Bytes + Relative_Offset_Field_Sizes_In_Bytes;
+    public static final int Blob_Relative_Offset_Field_Offset_In_Bytes =
+        UserMetadata_Relative_Offset_Field_Offset_In_Bytes + Relative_Offset_Field_Sizes_In_Bytes;
+
+    // crc field start offset
+    public static final int Crc_Field_Offset_In_Bytes =
+        Blob_Relative_Offset_Field_Offset_In_Bytes + Relative_Offset_Field_Sizes_In_Bytes;
+
+    public static int getHeaderSize() {
```

Comment:
java docs

Reply:
Like I said, these are self-descriptive. MessageFormatRecord uses the convention of javadoc at the class level.